content/blog/2025-05-13-flat-bril/index.md

```diff
@@ -0,0 +1,170 @@
+(Benchmarks were run on a 2023 M4 Macbook Pro.)
+
+Rust Brili outperforms our flattened interpreter for the vast majority of benchmarks, although our interpreter has the lowest execution time out of the three interpreters for a few benchmarks (`call`, `call-with-args`, `mccarthy91`). 
+
+The benchmarks where there was the biggest discrepancy between the three interpreters’ execution time were programs with many recursive calls (`delannoy`, `ackermann`, `catalan`, `primes-between`). For these benchmarks, we notice that Rust Brili significantly outperforms both our interpreter & TypeScript Brili. Examining the source code for Rust Brili, we suspect this is because they maintain an explicit call stack data structure and associate each function with appropriate stack pointers, whereas when our interpreter handles function calls, we create brand new `HashMap`s to represent the local variables in each function, which is perhaps more expensive than their approach. 
+
+<div style="display: flex; justify-content: center; gap: 10px;">
+  <img src="bench_results.png" alt="" style="width: 100%;">
+</div>
+
+Besides comparing our interpreter with the other Brili implementations, we also used the Samply profiler to figure out where our interpreter was spending most of its time. The stack chart below (obtained from the Firefox Profiler UI) demonstrates how our interpreter exectuable spends its time on the benchmark `catalan.bril` – we chose to highlight this benchmark since it has many recursive function calls and the difference between our interpreter and the TypeScript/Rust Brili interpreters’ performance is noticeable. 
+
+The stack chart shows that most of the runtime of the executable spent was in the interpreter-related functions: `mmap`-ing the flat file  format was relatively quick, and so was converting the JSON to a flat file format (omitted from the screenshot). Zooming into the stack chart (second screenshot below) and focusing on the function calls towards the bottom, we see that when interpreting individual individual instructions, our interpreter calls a lot of standard library functions that manipulate `HashMap`s and `Vec`s. This is because the environment datatype in our interpreter is defined as `HashMap<&str, BrilValue>` (mapping variable names to a Bril value), i.e. keys in the hashmaps are (references to) strings. We realized that strings were the canonical way to represent variables in the environment, since different indexes in the arguments field of an instruction may point to the same underlying variable. However, the fast Rust Brili interpreter canonicalises variable names into a numerical representation (as described in their blogpost), allowing for faster look-ups in their environment. We suspect that this is one of the reasons why the Rust Brili interpreter out-performs our flat interpreter. 
```

Comment:
Thanks for doing this investigation! I agree that it is likely that this is the performance bottleneck here. It would be interesting to try to alleviate this bottleneck by requiring that variable name indices be unique…